src/mgr/ActivePyModules.cc

```diff
@@ -639,7 +639,7 @@ void ActivePyModules::set_store(const std::string &module_name,
 void ActivePyModules::set_config(const std::string &module_name,
     const std::string &key, const boost::optional<std::string>& val)
 {
-  module_config.set_config(&monc, module_name, key, val);
+  module_config.set_config(&monc, module_name, key, val, true);
```

Comment:
@liewegas this really looks awkward...

Do I understand it right that the problem is that the `val` pointer should be protected by GIL when processed by `PyFormatter`?
What if we made `module_config.set_config` be always called with GIL dropped. And in the cases like this, where `val` should be protected by GIL, we copy `val` to a new string variable before dropping the GIL and calling `module_config.set_config`?

Reply:
We could change this func signature to `void ActivePyModules::set_config(const std::string &module_name, const std::string &key, boost::optional<std::string> val)` and then could safely drop the GIL before calling `module_config.set_config`.

Reply:
And probaly `key` should be copied too.

Reply:
I think val and key are stable because the caller has a reference to the underlying py object.  The problem is that PyFormatter is creating new python objects and that can't be done without the GIL without corrupting cpython's internals.

Maybe a cleaner approach is to have the caller drop the gil, and have these methods retake GIL explicitly while using PyFormatter?

Reply:
(BTW I tried to add a call to a Py_something method so that we can assert that we hold GIL in PyFOrmatter and elsewhere but the method doesn't work (always returns true) when using submodules.  Meh.)

Reply:
> Maybe a cleaner approach is to have the caller drop the gil, and have these methods retake GIL explicitly while using PyFormatter?

It looks cleaner for me too.